### PR TITLE
[core] Introducing InstrumentedIOContextWithThread.

### PR DESCRIPTION
**Pull request by rynewang**
Previously we had several ad-hoc places to do a "thread and io_context" pattern: create a thread dedicated to an asio io_context, then workload can post async tasks onto it. This makes duplicate code: everywhere we create threads, implement stop and join.

Introducing InstrumentedIOContextWithThread that does exactly this and replaces existing usages.

Also fixes some absl::Time computations with best practice.

This is refactoring. Should have no runtime difference.

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 1]
```diff
@@ -264,14 +267,10 @@ void GcsServer::DoStart(const GcsInitData &gcs_init_data) {
 void GcsServer::Stop() {
   if (!is_stopped_) {
     RAY_LOG(INFO) << "Stopping GCS server.";
-    ray_syncer_io_context_.stop();
-    ray_syncer_thread_->join();
-    ray_syncer_.reset();
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_server.cc:269]: we should keep this line?

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 2]
```diff
@@ -264,14 +267,10 @@ void GcsServer::DoStart(const GcsInitData &gcs_init_data) {
...
 
-    gcs_task_manager_->Stop();
-
-    pubsub_handler_->Stop();
-    pubsub_handler_.reset();
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_server.cc:274]: same here

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 3]
```diff
@@ -54,6 +54,9 @@ GcsServer::GcsServer(const ray::gcs::GcsServerConfig &config,
     : config_(config),
       storage_type_(GetStorageType()),
       main_service_(main_service),
+      pubsub_io_context_("pubsub_io_context"),
+      task_io_context_("task_io_context"),
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_server.cc:58]: I think it's better for GcsTaskManager to own and manage its io_context like before?

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 4]
```diff
@@ -54,6 +54,9 @@ GcsServer::GcsServer(const ray::gcs::GcsServerConfig &config,
     : config_(config),
       storage_type_(GetStorageType()),
       main_service_(main_service),
+      pubsub_io_context_("pubsub_io_context"),
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_server.cc:57]: InternalPubSubHandler should own it?

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 3 above)
Comment (rynewang) [reply src/ray/gcs/gcs_server/gcs_server.cc:58]: why do we need that? I think all managers can share a same style.

**Review by jjyao — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 4 above)
Comment (rynewang) [reply src/ray/gcs/gcs_server/gcs_server.cc:57]: per offline discussion we keep it in gcs server

**Pull request merged**